File: aoc2024/src/ts/day19.ts
import { Grid, linesToCharGrid, toNumberGrid } from "./utils/char-grid";
import { readInput } from "./utils/file-utils";
import { CardinalDirection, Point } from "./utils/point";

const lines = readInput("day19", false);

function isPossible(design: string, towels: string[]): boolean {
  const candidates = towels.filter((t) => design.startsWith(t));
  for (let i = 0; i < candidates.length; i++) {
    const t = candidates[i];
    if (design.length === t.length) {
      return true;
    }
    if (isPossible(design.slice(t.length), towels)) {
      return true;
    }
  }
  return false;
}

const cache = new Map<string, number>();

function getArrangementCount(design: string, towels: string[]): number {
  if (cache.has(design)) {
    // console.log(`cache hit for ${design}`);
    return cache.get(design)!;
  }

  const candidates = towels.filter((t) => design.startsWith(t));
  let results = 0;
  //   console.log(`${design} : checking ${candidates}`);
  for (let i = 0; i < candidates.length; i++) {
    const t = candidates[i];
    if (design.length === t.length) {
      results += 1;
    } else {
      const a = getArrangementCount(design.slice(t.length), towels);
      results += a;
    }
  }

  cache.set(design, results);
  return results;
}

function part1() {
  const towels = lines[0].split(", ");
  const designs = lines.slice(2);

  let count = 0;
  designs.forEach((design) => {
    if (isPossible(design, towels)) {
      count += 1;
    }
  });

  console.log(`Count ${count}`);
}

function part2() {
  const towels = lines[0].split(", ");
  const designs = lines.slice(2);

  let count = 0;
  designs.forEach((design) => {
    const a = getArrangementCount(design, towels);
    console.log(`*** ${design} : ${a}\n`);
    count += a;
  });

  console.log(`Count ${count}`);
}

part2();

File: aoc2024/src/python/day19.py
from collections import defaultdict
import time
from typing import Set, Tuple, List
from aoc_utils import  A_star, manhattan_distance, splitInts, runIt, PuzzleInput, split_on_empty_lines, sub, Grid, Point, cardinal_directions, add, North, South, East, West
import functools
import math
import re
from itertools import chain, combinations, permutations, product
import sys
import numpy


testInput = r"""r, wr, b, g, bwu, rb, gb, br

brwrr
bggr
gbbr
rrbgbr
ubwu
bwurrg
brgr
bbrgwb"""

input = PuzzleInput('input/day19.txt', testInput)
lines = input.getInputLines(test=False)

def is_possible(towel: str, patterns: list[str]) -> bool:
    for p in [p for p in patterns if towel.startswith(p)]:
        if len(towel) == len(p) or is_possible(towel[len(p):], patterns):
            return True
    return False

def count_arrangements(towel: str, patterns: list[str], counts: dict[str, int]) -> int:
    if towel in counts:
        return counts[towel]
    
    c = 0
    for p in [p for p in patterns if towel.startswith(p)]:
        if len(towel) == len(p):
            c += 1
        else:
            c += count_arrangements(towel[len(p):], patterns, counts)
    
    counts[towel] = c
    return c


def part1():
    patterns = list(lines[0].split(', '))
    patterns.sort(key=len)
    patterns.reverse()
    towels = lines[2:]

    c = sum(1 for t in towels if is_possible(t, patterns))
    print(f'Possible count {c}')


def part2():
    patterns = lines[0].split(', ')
    towels = lines[2:]

    counts = {}
    c = sum(count_arrangements(t, patterns, counts) for t in towels)
    print(f'Arrangements count {c}')
    
runIt(part1, part2)
